File: trees/check_subtree.py
#!/bin/python

from trees.preorder import pre_order

def check_subtree(root_1, root_2):
    node = root_2
    while node:
        if node == root_1:
            return True
        node = node.parent
    return False


def check_subtree_2(root_1, root_2):
    Open = [root_1]
    while Open:
        node = Open.pop(0)
        if node == root_2:
            return True
        if node.right:
            Open.append(node.right)
        if node.left:
            Open.append(node.left)
    return False


def check_subtree_3(root_1, root_2):
    str_1 = pre_order(root_1)
    str_2 = pre_order(root_2)
    return str_2 in str_1

File: cracking_the_coding_interview_book/chapter_4/check_subtree.cpp
#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node* right;
        Node* left;

        Node(int value) {
            data = value;
            right = NULL;
            left = NULL;
        }
};

bool match_tree(Node* bt_1, Node* bt_2) {
    if (bt_1 == NULL && bt_2 == NULL)
        return true;
    if (bt_1 == NULL || bt_2 == NULL)
        return false;
    if (bt_1->data != bt_2->data)
        return false;
    return (match_tree(bt_1->left, bt_2->left) && match_tree(bt_1->right, bt_2->right));
}

bool check_subtree(Node* bt_1, Node* bt_2) {
    if (bt_1 == NULL)
        return false;
    if (match_tree(bt_1, bt_2))
        return true;
    else
        return (check_subtree(bt_1->left, bt_2) || check_subtree(bt_1->right, bt_2));
}

bool contains(Node* bt_1, Node* bt_2) {
    if (bt_2 == NULL) return true;
    return check_subtree(bt_1, bt_2);
}
